Some improvements

L117 opened the pull request:
Short summary:

- Config file format changed to TOML.
- Updated `uptime` command to work with `serenity`'s `ShareMap` instead of `static`.
- Implemented `info` command ~~(more or less, join time still needs to be properly formatted)~~.
- Removed on-join functionality as in python version of this bot.
- Implemented `permissions` command.

Commit:
Cargo.lock is not supposed to be in repo.

Commit:
Removed member join event as it was decided to remove it cuz it only annoys people.

Commit:
Changed config file format to TOML as TOML is de-facto standard in a Rust world.

Commit:
Added `memberinfo` command.

Commit:
Updated `uptime` command to work with `serenity`'s ShareMap instead on nasty static.

Commit:
It seems that binaries should have Cargo.lock in VCS.
http://doc.crates.io/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries
Returning it back.

zeyla commented on src/uptimer.rs:
```diff
@@ -0,0 +1,21 @@
+impl Uptimer {
+    pub fn new() -> Uptimer {
+        Uptimer {
+            started_at: now_utc()
```
It's common practice to not bring the function itself into scope, and to instead `use time;` and then `time::now_utc()`.

zeyla commented on src/commands/meta.rs:
```diff
@@ -1,44 +1,96 @@
+                    roles.push_str(", ");
+                }
+            }
+            Some((roles, member.joined_at.clone()))
```
This clone feels unnecessary -- you should be able to create a new `CreateEmbed` builder, moving the footer call here, and then returning it to successfully avoid the clone.

zeyla commented on src/commands/meta.rs:
```diff
@@ -1,44 +1,96 @@
+            .field(|cef| 
+                cef.inline(true)
+                .name("Id")
+                .value(&format!("{}", id))
```
Prefer `id.to_string()` over formatting `{}`.

zeyla commented on src/commands/meta.rs:
```diff
@@ -1,44 +1,96 @@
+    let ref name = msg.author.name;
+    let dtag = msg.author.distinct();
+    let created = at_utc(msg.author.created_at());
+    let created = created.asctime();
```
The values bound to all of these are only used in the matching arm of the if-let below, this should be moved within it so that these aren't unnecessarily calculated.

zeyla commented on Cargo.toml:
```diff
@@ -6,10 +6,16 @@ authors = [
 
+[dependencies.serenity]
+version = "0.2"
+features = ["cache"]
```
This only needs to be specified if you're not using default features. serenity enables the `cache` and `framework` features in v0.2.0 by-default.

zeyla commented on src/commands/meta.rs:
```diff
@@ -1,44 +1,96 @@
+                if let Some(role) = role_id.find() {
+                    roles.push_str(&role.name);
+                } else {
+                    return None;
```
Are you sure you don't mean to `continue;` here?

L117 replied:
i was unsure. What does this actually mean, when there's `role_id` but no `role` for it? Sound like case where something went wrong, so code goes to else branch in `if let` below displaying error.

L117 commented on src/commands/meta.rs:
```diff
@@ -1,44 +1,96 @@
+
+command!(memberinfo(_context, msg, _args) {
+    let avatar_url = msg.author.avatar_url().or_else(||
+        msg.author.static_avatar_url()
```
And there's also `get_face` you've added recently. I was going to replace this in next commits.

Commit:
Some code updates according to advices.

Commit:
Unused code/imports removal. Clippy lints.

Commit:
Added color to embeds.

Commit:
Update to `serenity`'s `master` and some minor changes.

Commit:
Added `permissions` command. Renamed `memberinfo` command into just `info`.

Commit:
Fix for `permissions` command.

Commit:
Fix for `info` role.

As embed fields must contain non-empty names and values it was failing when member has no roles except `@everyone`.

Commit:
Replaced that ugly permissions formatting macro invocation with something less ugly.

ivandardi commented:
Is this pull request all finished? Because if it is, just give me a heads up and I'll merge.

L117 commented:
Actually I was going to implement `remind` to close #3. But I can create another PR for this.

ivandardi commented on src/commands/utilities.rs:
```diff
@@ -21,15 +21,20 @@ pub async fn go(ctx: Context<'_>) -> Result<(), Error> {
+	discard_spare_arguments
+)]
 pub async fn source(ctx: Context<'_>) -> Result<(), Error> {
 	ctx.say("https://github.com/rust-community-discord/rustbot")
```
If you could change this to https://github.com/rust-community-discord/ferrisbot-for-discord, it'd be much appreciated.

danielrab replied:
ran rustfmt and replaced the link.
